chore(351): add service reliability standard (#352)
UKHomeOffice/engineering-guidance-and-standards: docs/standards/service-reliability.md

---
layout: standard
order: 1
title: Service reliability

> date: 2024-01-12

Reviewer: ```suggestion
date: 2024-04-26
```

> id: SEGAS-00013

Reviewer: The new developer-testing standards is SEGAS-00013, this will need to be SEGAS-00014 now

  - [Service MUST be tolerant to expected and unexpected failure of upstream services such as databases](#service-must-be-tolerant-to-expected-and-unexpected-failure-of-upstream-services-such-as-databases)
  - [Service MUST be sized appropriately for normal operations and MUST be able to automatically scale as appropriate based on metrics](#service-must-be-sized-appropriately-for-normal-operations-and-must-be-able-to-automatically-scale-as-appropriate-based-on-metrics)
  - [Service MUST be able to handle requests to protect the overall service](#service-must-be-able-to-handle-requests-to-protect-the-overall-service)
  - [Service MUST be able to tolerate expected load](#service-must-be-able-to-tolerate-expected-load)
  - [Service MUST be able to tolerate expected stress](#service-must-be-able-to-tolerate-expected-stress)
  - [Service MUST be soak tested](#service-must-be-soak-tested)

### Service reliability MUST be observable relative to defined service level expectations

> Reliability expectations should be clear and baselined so that Service Level Objectives (SLOs) can be defined for measures that can keep Home Office services reliable.

Reviewer 2: ```suggestion
Reliability expectations should be clear and baselined so that service level objectives (SLOs) can be defined for measures that can keep Home Office services reliable.
```

### Service MUST be tolerant to expected and unexpected failure of a data centre

In the event of an unexpected catastrophic event (flooding, fire, etc.) or an expected event (data centre maintenance), an entire data centre may be lost causing outages of services that are hosted in the affected data centre.

> Although rare, services should consider the possibility and guard against such scenarios by spreading applications / services across multiple data centres (e.g. Availability Zones) either in Active/Active or Active/Passive configuration.

Reviewer 2: ```suggestion
Although rare, services should consider the possibility and guard against such scenarios by spreading applications / services across multiple data centres (e.g. availability zones) either in active/active or active/passive configuration.
```

### Service MUST be tolerant to expected and unexpected failure of physical or virtualised hardware as well as network

> In the event of an unexpected event (hardware fault, withdrawal of virtualised hardware e.g. SPOT instances) or an expected event (maintenance activities) your service may suffer an outage.

Reviewer 2: ```suggestion
In the event of an unexpected event (hardware fault, withdrawal of virtualised hardware e.g. spot instances) or an expected event (maintenance activities) your service may suffer an outage.
```

> Such scenario is more common and your service should guard against such scenarios by operating;

Reviewer 2: ```suggestion
Such scenario is more common and your service should guard against such scenarios by operating:
```

- from multiple data centres
- from multiple physical / virtualised nodes

### Service MUST be tolerant to expected and unexpected failure of upstream services such as databases

In the event of an unexpected event (failure of an upstream application, database, etc.) or an expected event (upgrade, maintenance activities, etc.) your service may suffer an outage.

> Your service should guard against such scenarios by;
> - handling the failure gracefully, or
> - attempt to reconnect, and/or
> - attempt to retry within reasonable limits of your service

Reviewer 2: ```suggestion
Your service should guard against such scenarios by:
- handling the failure gracefully
- attempting to reconnect
- attempting to retry within reasonable limits of your service
```

> When retrying transactions, particularly when an upstream service was down, avoid flooding the upstream service(s) with requests thus potentially creating a DDoS style attack on the upstream service, potentially causing further upstream service outages. 
> Furthermore, implement a backoff strategy if your transactions allow it.

Reviewer 2: ```suggestion
When retrying transactions, particularly when an upstream service was down, avoid flooding the upstream service(s) with requests thus potentially creating a DDoS style attack on the upstream service, potentially causing further upstream service outages. Furthermore, implement a backoff strategy if your transactions allow it.
```

Ideally, scaling should be based on some metric that is indicative of an application's performance (such as number of busy threads, queue depth) rather than system resources such as CPU and memory. Not scaling based on intelligent metrics could result in situations such as:
- your application has exhausted threads and cannot service any more requests, while CPU and memory utilisation is low
- a messaging queue has filled up, and there are not enough consumers to process messages

Basing scaling policies on system resources should be the fallback option, rather than the main criteria by which you scale.
Allowing for automatic scaling of a service not only enables a service to respond accordingly during peak or off-peak times, but is also a valuable cost saver.

**Services should be continuously reviewed.**

### Service MUST be able to handle requests to protect the overall service

The service must be able to apply rate limiting rules to protect itself from being overwhelmed by external and internal clients, preventing such scenarios as DDoS style attacks either intentional or accidental.

> - Check that services respond accordingly during load and stress testing; usually services should respond with HTTP 503

Reviewer 2: ```suggestion
Check that services respond accordingly during load and stress testing; usually services should respond with HTTP 503
```

### Service MUST be able to tolerate expected load

> Where required based off business requirements, the service can be load, soak or stressed tested to expected normal conditions as defined by the service volumetrics to; 

Reviewer 2: ```suggestion
Where required based off business requirements, the service can be load, soak or stressed tested to expected normal conditions as defined by the service volumetrics to: 
```

- ensure the application can manage expected load
- identify and remediate performance bottlenecks
- understand trigger points for scaling policies

### Service MUST be able to tolerate expected stress

> The service must be stressed tested beyond normal load conditions (at least 200% of expected peak load), so as to;

Reviewer 2: ```suggestion
The service must be stressed tested beyond normal load conditions (at least 200% of expected peak load), so as to:
```

> - Understand how the application behaves beyond normal conditions, and document any findings
> - Ensure scaling policies behave as expected, and that normal service can be expected from unexpected load conditions
> - Identify and remediate performance bottlenecks in the service and upstream / downstream systems and services.
> - This is particularly important as there will be hidden limits that affect scaling capacity, such as limits on upstream systems, Cloud accounts, or network throughput on appliances connecting to 3rd parties/POISE.

Reviewer 2: ```suggestion
- understand how the application behaves beyond normal conditions, and document any findings
- ensure scaling policies behave as expected, and that normal service can be expected from unexpected load conditions
- identify and remediate performance bottlenecks in the service and upstream / downstream systems and services

This is particularly important as there will be hidden limits that affect scaling capacity, such as limits on upstream systems, cloud accounts, or network throughput on appliances connecting to 3rd parties/POISE.
```

### Service MUST be soak tested

The service must be soak tested to ensure normal operating conditions over time, from low, normal to high load situations, so as to:

> - Understand how the service behaves over a period of time, and that garbage collection effectively prevents memory and CPU leaks over time
> - Understand how the service recovers from high load situations and ensure it relinquishes unused resource back to the underlying infrastructure

Reviewer 2: ```suggestion
- understand how the service behaves over a period of time, and that garbage collection effectively prevents memory and CPU leaks over time
- understand how the service recovers from high load situations and ensure it relinquishes unused resource back to the underlying infrastructure
```

> ---

Reviewer 2: ```suggestion
```

In the event of an unexpected event (hardware fault, withdrawal of virtualised hardware e.g. SPOT instances) or an expected event (maintenance activities) your service may suffer an outage.
Such scenario is more common and your service should guard against such scenarios by operating;
- from multiple data centres
- from multiple physical / virtualised nodes

### Service MUST be tolerant to expected and unexpected failure of upstream services such as databases

In the event of an unexpected event (failure of an upstream application, database, etc.) or an expected event (upgrade, maintenance activities, etc.) your service may suffer an outage.
Your service should guard against such scenarios by;
- handling the failure gracefully, or
- attempt to reconnect, and/or
- attempt to retry within reasonable limits of your service

When retrying transactions, particularly when an upstream service was down, avoid flooding the upstream service(s) with requests thus potentially creating a DDoS style attack on the upstream service, potentially causing further upstream service outages. 
Furthermore, implement a backoff strategy if your transactions allow it.

> Go through upstream services your service relies on, and test how it reacts to an expected or unexpected failure of that service.

Reviewer 3: This is more testing guidance, this will probably not be needed.

> Go through upstream services your service relies on, and test how it reacts to an expected or unexpected failure of that service.

Author: ```suggestion
```

### Service MUST be sized appropriately for normal operations and MUST be able to automatically scale as appropriate based on metrics

The service must be sized appropriately in terms of system resources (CPU, memory, storage, etc.) for normal operating conditions preventing unnecessary resource wastage.

All components of a service can be scaled up or out (or down or in) based on some metric (such as CPU, memory, thread utilisation or some appropriate metric) so as to ensure continued successful operation of the service during expected and unexpected peak or off-peak operating periods.

Ideally, scaling should be based on some metric that is indicative of an application's performance (such as number of busy threads, queue depth) rather than system resources such as CPU and memory. Not scaling based on intelligent metrics could result in situations such as:
- your application has exhausted threads and cannot service any more requests, while CPU and memory utilisation is low
- a messaging queue has filled up, and there are not enough consumers to process messages

> Basing scaling policies based on system resources should be the fallback option, rather than the main criteria by which you scale.

Reviewer 3: I think 'based' isn't needed in this sentence.

> Basing scaling policies based on system resources should be the fallback option, rather than the main criteria by which you scale.

Author: ```suggestion
Basing scaling policies on system resources should be the fallback option, rather than the main criteria by which you scale.
```

The service must be able to apply rate limiting rules to protect itself from being overwhelmed by external and internal clients, preventing such scenarios as DDoS style attacks either intentional or accidental.

- Check that services respond accordingly during load and stress testing; usually services should respond with HTTP 503

### Service MUST be able to tolerate expected load

Where required based off business requirements, the service can be load, soak or stressed tested to expected normal conditions as defined by the service volumetrics to; 

- ensure the application can manage expected load
- identify and remediate performance bottlenecks
- understand trigger points for scaling policies

### Service MUST be able to tolerate expected stress

The service must be stressed tested beyond normal load conditions (at least 200% of expected peak load), so as to;

- Understand how the application behaves beyond normal conditions, and document any findings
- Ensure scaling policies behave as expected, and that normal service can be expected from unexpected load conditions
- Identify and remediate performance bottlenecks in the service and upstream / downstream systems and services.

>   - This is particularly important as there will be hidden limits that affect scaling capacity, such as limits on upstream systems, Cloud accounts, or network throughput on appliances connecting to 3rd parties/POISE.

Author: ```suggestion
- This is particularly important as there will be hidden limits that affect scaling capacity, such as limits on upstream systems, Cloud accounts, or network throughput on appliances connecting to 3rd parties/POISE.
```

>   - This is particularly important as there will be hidden limits that affect scaling capacity, such as limits on upstream systems, Cloud accounts, or network throughput on appliances connecting to 3rd parties/POISE.

Reviewer 3: I think there's an extra space at the beginning of this line. I also think that 'appliances' should say 'applications'

  - [Service MUST be soak tested.](#service-must-be-soak-tested)

### Service reliability MUST be observable relative to defined service level expectations

Reliability expectations should be clear and baselined so that Service Level Objectives (SLOs) can be defined for measures that can keep Home Office services reliable.

### Service MUST be tolerant to expected and unexpected failure of a data centre

In the event of an unexpected catastrophic event (flooding, fire, etc.) or an expected event (data centre maintenance), an entire data centre may be lost causing outages of services that are hosted in the affected data centre.

Although rare, services should consider the possibility and guard against such scenarios by operating from multiple data centres by spreading applications / services across multiple data centres (e.g. AWS Availability Zones) either in Active/Active or Active/Passive configuration.

Check that services are spread across multiple nodes and/or multiple availability zones; this can be checked by checking;
- Kubernetes pod anti-affinity rules,
- AWS multi-az settings for AWS services

> - AWS Auto Scaling Group configurations

Author: removed checks

---
layout: standard
order: 1
title: Service reliability
date: 2023-11-23
id: SEGAS-00013
tags:
  - Reliability
  - Infrastructure
  - SRE
---

> Home office systems should be able to withstand failures during various circumstances, both expected and unexpected over a specified period in specified environments. 

Reviewer 4: The end of this sentence seems a little vague? Surely if it's unexpected, it's not over a specified period?

Author: @aaronrussellHO How would you put it otherwise? - (to make the point that systems should be able to cope with all eventualities)

Reviewer 4: ```suggestion
Home Office systems should be able to withstand different eventualities, this could include scheduled maintenance, data centre failure and increased load and stress.``

Reviewer: I don't think we want to lose the essence of the point around failures over a specified period of time. We want people to build reliability to meet sensibly defined expected service levels.

Reviewer 5: I'm a bit worried that if we focus too much on service levels (SLOs/SLAs?) then they might be so easy to meet that we don't do some of the common sense stuff. (You see this with performance testing, for example.)

I'd think that if we build things to be as reliable as possible, we'll probably find that we far exceed the required service level.

With regards to this paragraph, I'm a little confused as to what it's getting at. Is this about being resilient in the face of any failure modes that can be anticipated? (Which requires taking the time to think about what those failure modes are.)

Reviewer: Agree that we don't want to under or over engineer for reliability (though that will always happen to an extent). I think that is why we want to talk a little bit about service levels, to put sensible bounds and direction for what is required, discouraging the unnecessary.

We can probably hold off fully re-writing the opening description for the standard until the requirements are more concrete, the description will usually just summarise the whole. I'd say the intent is for the standard to call out the minimum expectations, that apply for any service at HO, with scope for people to extend dependant on their case.

  - [Service MUST be able to set appropriate caching directives in HTTP responses so as to reduce load on the service](#service-must-be-able-to-set-appropriate-caching-directives-in-http-responses-so-as-to-reduce-load-on-the-service)
  - [Service MUST be able to toletate expected load](#service-must-be-able-to-toletate-expected-load)
  - [Service MUST be able to toletate expected stress](#service-must-be-able-to-toletate-expected-stress)
  - [Service MUST be soak tested.](#service-must-be-soak-tested)

### Service reliability MUST be observable relative to defined service level expectations
Reliability expectations should be clear and baselined so that Service Level Objectives (SLOs) can be defined for measures that can keep Home Office services reliable.

### Service MUST be tolerant to expected and unexpected failure of a data centre

In the event of an unexpected catastrophic event (flooding, fire, etc.) or an expected event (data centre maintenance), an entire data centre may be lost causing outages of services that are hosted in the affected data centre.

> Although rare, services should consider the possibility and guard against such scenarios by operating from multiple data centres by spreading applications / services across multiple data centres (e.g. AWS Availability Zones) either in Active/Active or Active/Passive configuration.

Reviewer 3: "by operating from multiple data centres" feels like a duplicate clause and can probably be removed.

Author: removed checks

> Check that services are spread across multiple nodes and/or multiple availability zones; this can be checked by checking;

Reviewer 3: "this can be met with" is a less clunky clause then "this can be checked by checking;"

Author: removed checks

- Kubernetes pod anti-affinity rules,
- AWS multi-az settings for AWS services
- AWS Auto Scaling Group configurations

Evidence could be a screenshot, link to repository or code snippet of where it is configured. 
**Example:** [Kubernetes Docs on pod anti-affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#an-example-of-a-pod-that-uses-pod-affinity:~:text=Pod%20affinity%20example)

### Service MUST be tolerant to expected and unexpected failure of physical or virtualised hardware, including networks

In the event of an unexpected event (hardware fault, withdrawal of virtualised hardware e.g. SPOT instances) or an expected event (maintenance activities) your service may suffer an outage.
Such scenario is more common and your service should guard against such scenarios by operating;
- from multiple data centres
- from multiple physical / virtualised nodes

> Check that services are spread across multiple nodes and/or multiple availability zones; this can be checked by checking;

Reviewer 3: Same comment as for the above Standard, I'm not sure we need lines 56-59 as this was originally for QAT testing purposes.

Author: removed checks

- AWS Auto Scaling Group configurations

Evidence could be a screenshot, link to repository or code snippet of where it is configured.
**Example:** [Kubernetes Docs on pod anti-affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#an-example-of-a-pod-that-uses-pod-affinity:~:text=Pod%20affinity%20example)

### Service MUST be tolerant to expected and unexpected failure of upstream services such as databases

In the event of an unexpected event (failure of an upstream application, database, etc.) or an expected event (upgrade, maintenance activities, etc.) your service may suffer an outage.
Such scenarios are common and your service should guard against such scenarios by;
- handling the failure gracefully, or
- attempt to reconnect, and/or
- attempt to retry within reasonable limits of your service

When retrying transactions, particularly when an upstream service was down, avoid flooding the upstream service(s) with requests thus potentially creating a DDoS style attack on the upstream service, potentially causing further upstream service outages. 
Furthermore, implement a backoff strategy if your transactions allow it.

> - Go through upstream services your service relies on, and  how it reacts to an expected or unexpected failure of that service.

Reviewer 3: "Feel as if we need another word in there, maybe "Go through upstream services your service relies on, and **test** how it reacts"

Author: removed checks

> - Tests results for service proving behaviour.

Reviewer 3: Assumes external test team, could be removed.

> ### Service MUST be sized appropriately for normal operations

Reviewer: I think the spirit of this requirement could be split out to supplement some of the other requirements. 

e.g.
- definition of 'normal operations' probably belongs in first requirement on definition of SLOs, along with regular reviews and setting acceptable mean time to recover
- sizing and scaling are heavily linked - initial sizing choices must enable appropriate scaling strategies for meeting load and cost optimisation (cost optimisation will be a different standard at some point I'd expect)
- assurance of the appropriateness of these things is part of the load, stress and soak testing requirements

> ### Service MUST be sized appropriately for normal operations

Author: Following discussion at Service Reliability GUILD

Merged point on sizing with point on autoscaling.

Will resolve this comment, but do feel free to raise another comment on the new status on the PR.

Cheers.

The application / service must be sized appropriately in terms of system resources (CPU, memory, storage, et al) for normal operating conditions preventing unnecessary resource wastage. 

> Considering the acceptable time it takes for your application to recover (Mean Time To Recover / MTTR), additional headroom is allowed in the event of failures - for example, it is permissible if you are running a two node cluster that normal operations yield 50% resource utilisation, so that if a single node is lost your service may continue to function on a single node at 100% utilisation until the second node is/can be recovered.

Reviewer 5: Assuming this lives somewhere, is it worth calling out auto-scaling here?

Reviewer 5: Oh sorry, that's the next section.

- Check that Kubernetes services have configured CPU and Memory requests and limits are configured

> - Check that AWS services are sized appropriately

Reviewer 3: Do we want to go tech agnostic and just refer to cloud or infrastructure services?

> - Refer to appropriate dashboards to ensure CPU and Memory utilisation. 

Reviewer 3: Should probably be "Refer to appropriate dashboards to ensure CPU and Memory utilisation **are sized appropriately"**

**Services should be continuously reviewed.**

### Service MUST be able to scale as appropriate automatically based on metrics

All components of a service can be scaled up or out (or down or in) based on some metric (such as CPU, memory, thread utilisation or some appropriate metric) so as to ensure continued successful operation of the service during expected and unexpected peak or off-peak operating periods.

> Ideally, scaling should be based on some metric that is indicative of an application's performance (such as number of busy threads, queue depth) rather than system resources such as CPU and memory as (in the case of scaling up for instance);

Reviewer 3: This wording has always been hard for me to read - I would rephrase to

Ideally, scaling should be based on some metric that is indicative of an application's performance (such as number of busy threads, queue depth) rather than system resources such as CPU and memory. Not scaling based on intelligent metrics could result in situations such as:
- your application has exhausted threads and cannot service any more requests, while CPU and memory utilisation is low 
- a messaging queue has filled up, but there are not enough consumers to process messages

Basing scaling policies based on system resources should be the fallback option, rather than the main criteria by which you scale.
Allowing for automatic scaling of a service not only enables a service to respond accordingly during peak or off-peak times, but is also a valuable cost saver.

- Check that Kubernetes service have HPA configured
- Check that AWS services are in Auto-Scaling Groups

### Service MUST be able to reject and/or queue requests to protect the overall service

The service must be able to apply rate limiting rules to protect itself from being overwhelmed by external and internal clients, preventing such scenarios as DDoS style attacks either intentional or accidental.

- Check that services respond accordingly during load and stress testing; usually services should respond with HTTP 503
- Tests results for service proving behaviour.

### Service MUST be able to set appropriate caching directives in HTTP responses so as to reduce load on the service

Where possible, the service should set caching headers to HTTP requests to reduce duplicate computation and thereby reducing load, conversely, client services should cache responses as defined by HTTP responses.

> - Check that services respond with appropriate cache directives in HTTP response headers

Reviewer 5: We often can't allow caching in order to protect our data. I agree we should do it for non-private static assets though. I know you've got 'where possible', but if we open with that is it appropriate to have this as a MUST rather than a SHOULD?

Reviewer 4: Yeah, I agree, I think this should call out static assets in particular, and any public endpoints

Reviewer 6: think that "appropriate" needs to be made clear here.

Author: Removed requirement altogether from Standard.

> ### Service MUST be able to toletate expected load

Reviewer 5: ```suggestion
### Service MUST be able to tolerate expected load
```

Where required based off business requirements, the service can be load, soak or stressed tested to expected normal conditions as defined by the service volumetrics to; 

- ensure the application can manage expected load
- identify and remediate performance bottlenecks
- understand trigger points for scaling policies

Validate with test teams that the service has been tested to expected peak load.
Provide evidence / links to test reports.

> ### Service MUST be able to toletate expected stress

Reviewer 5: ```suggestion
### Service MUST be able to tolerate expected stress
```

> The service must be stressed tested beyond normal load conditions (at least 200% of expected peak load), so as to;
> - Understand how the application behaves beyond normal conditions, and document any findings
> - Ensure scaling policies behave as expected, and that normal service can be expected from unexpected load conditions
> - Identify and remediate performance bottlenecks in the service and upstream / downstream systems and services.

Reviewer 4: ```suggestion
The service must be stressed tested beyond normal load conditions (at least 200% of expected peak load), so as to;

- Understand how the application behaves beyond normal conditions, and document any findings
- Ensure scaling policies behave as expected, and that normal service can be expected from unexpected load conditions
- Identify and remediate performance bottlenecks in the service and upstream / downstream systems and services.
```

Author: @aaronrussellHO being a big block of text, it's hard to see, is your suggestion mainly space and line break changes? or am I missing something? - a comment to go with suggested changes would be ideal.

Reviewer 4: Sorry, yes, just line spacing in this one.

  - This is particularly important as there will be hidden limits that affect scaling capacity, such as limits on upstream systems, Kubernetes, our AWS accounts, or network throughput on appliances connecting to 3rd parties/POISE.

> Validate with test teams that the service has been stress tested to expected peak load.

Reviewer 5: This assumes a separate test team, which I don't think is ideal.

Author: Following discussion at Service Reliability GUILD

Removed line.

Provide evidence / links to load test reports.

> ### Service MUST be soak tested.
> The service must be soak tested to ensure normal operating conditions over time, from low, normal to high load situations, so as to:
> - Understand how the service behaves over a period of time, and that garbage collection effectively prevents memory and CPU leaks over time
> - Understand how the service recovers from high load situations and ensure it relinquishes unused resource back to the underlying infrastructure

Reviewer 4: ```suggestion
### Service MUST be soak tested.

The service must be soak tested to ensure normal operating conditions over time, from low, normal to high load situations, so as to:

- Understand how the service behaves over a period of time, and that garbage collection effectively prevents memory and CPU leaks over time
- Understand how the service recovers from high load situations and ensure it relinquishes unused resource back to the underlying infrastructure
```

Author: @aaronrussellHO being a big block of text, it's hard to see, is your suggestion mainly space and line break changes? or am I missing something? - a comment to go with suggested changes would be ideal.

Reviewer 4: Same here, just line changes, no text changes

> Validate with test teams that the service has been soak tested to expected peak load.

Reviewer 5: This assumes a separate test team, which I don't think is ideal.

Author: Following discussion at Service Reliability GUILD

Removed line.

  - [Service MUST be tolerant to expected and unexpected failure of upstream services such as databases](#service-must-be-tolerant-to-expected-and-unexpected-failure-of-upstream-services-such-as-databases)
  - [Service MUST be sized appropriately for normal operations](#service-must-be-sized-appropriately-for-normal-operations)
  - [Service MUST be able to scale as appropriate automatically based on metrics](#service-must-be-able-to-scale-as-appropriate-automatically-based-on-metrics)
  - [Service MUST be able to reject and/or queue requests to protect the overall service](#service-must-be-able-to-reject-andor-queue-requests-to-protect-the-overall-service)
  - [Service MUST be able to set appropriate caching directives in HTTP responses so as to reduce load on the service](#service-must-be-able-to-set-appropriate-caching-directives-in-http-responses-so-as-to-reduce-load-on-the-service)
  - [Service MUST be able to toletate expected load](#service-must-be-able-to-toletate-expected-load)
  - [Service MUST be able to toletate expected stress](#service-must-be-able-to-toletate-expected-stress)
  - [Service MUST be soak tested.](#service-must-be-soak-tested)

> ### Service reliability MUST be observable relative to defined service level expectations
> Reliability expectations should be clear and baselined so that Service Level Objectives (SLOs) can be defined for measures that can keep Home Office services reliable.

Reviewer 4: ```suggestion
### Service reliability MUST be observable relative to defined service level expectations

Reliability expectations should be clear and baselined so that Service Level Objectives (SLOs) can be defined for measures that can keep Home Office services reliable.
```
Same here, spacing between heading and body

---
layout: standard
order: 1
title: Service reliability
date: 2023-11-23
id: SEGAS-00013
tags:
  - Reliability
  - Infrastructure
  - SRE
---

Home office systems should be able to withstand failures during various circumstances, both expected and unexpected over a specified period in specified environments. 

---

## Requirement(s)

> - [Requirement(s)](#requirements)

Reviewer: Don't need the first line here

```suggestion

```
